File: cpp/leetcode/leetcode/27_remove_element.cpp
//
//  remove_element.cpp
//  leetcode
//
//  Created by hatlonely on 16/1/13.
//  Copyright © 2016年 hatlonely. All rights reserved.
//
//  Given an array and a value, remove all instances of that value
//  in place and return the new length.
//
//  The order of elements can be changed. It doesn't matter what you
//  leave beyond the new length.
//

#include <iostream>
#include <cassert>
#include <vector>

namespace remove_element {
    
class Solution {
public:
    int removeElement(std::vector<int> &nums, int val) {
        int idx1 = 0;
        int idx2 = 0;
        
        int size = (int)nums.size();
        while (idx2 < size) {
            if (nums[idx2] == val) {
                idx2++;
            } else {
                nums[idx1++] = nums[idx2++];
            }
        }
        
        return idx1;
    }
};
    
int main(int argc, const char *argv[]) {
    return 0;
}
    
}

File: go/src/main/27_remove_element.go
package main

import "fmt"

func removeElement(nums []int, val int) int {
	j := 0
	for i := 0; i < len(nums); i++ {
		if nums[i] != val {
			nums[j] = nums[i]
			j++
		}
	}
	return j
}

func main() {
	for _, unit := range []struct {
		nums   []int
		val    int
		length int
	}{
		{[]int{3, 2, 2, 3}, 3, 2},
	} {
		if length := removeElement(unit.nums, unit.val); length != unit.length {
			fmt.Printf("removeElement(%v, %v) = %v != %v\n", unit.nums, unit.val, length, unit.length)
		}
	}
}
